mmdet3d/datasets/pipelines/transforms_3d.py

```diff
@@ -697,3 +698,74 @@ def __repr__(self):
         repr_str += '(bbox_enlarge_range={})'.format(
             self.bbox_enlarge_range.tolist())
         return repr_str
+
+
+@PIPELINES.register_module()
+class VoxelBasedPointSampler(object):
+    """Voxel based point sampler.
+
+    Apply voxel sampling to multiple sweep points.
+
+    Args:
+        name (str): Name of the dataset.
+        num_points (int): Number of points to be sampled.
+    """
+
+    def __init__(self, cur_sweep_cfg, prev_sweep_cfg=None):
+        self.cur_voxel_num = cur_sweep_cfg.pop('random_voxel_num')
+        self.cur_sweep_sampler = VoxelGenerator(**cur_sweep_cfg)
+        if prev_sweep_cfg is not None:
+            self.prev_voxel_num = prev_sweep_cfg.pop('random_voxel_num')
+            self.prev_sweep_sampler = VoxelGenerator(**prev_sweep_cfg)
+        else:
+            self.prev_voxel_num = 0
+
+    def __call__(self, results):
+        """Call function to sample points from multiple sweeps.
+
+        Args:
+            input_dict (dict): Result dict from loading pipeline.
+
+        Returns:
+            dict: Results after sampling, 'points', 'pts_instance_mask' \
+                and 'pts_semantic_mask' keys are updated in the result dict.
+        """
+        points = results['points']
+        cur_points_num = results['cur_points_num']
+        cur_sweep_points = points[:cur_points_num]
+        prev_sweeps_points = points[cur_points_num:]
+
+        voxels, coors, num_points_per_voxel = self.cur_sweep_sampler.generate(
+            cur_sweep_points)
+        max_voxels = self.cur_voxel_num
+        if voxels.shape[0] >= max_voxels:
+            cur_sweep_points = voxels[:max_voxels]
+        else:
+            choices = np.random.choice(
+                voxels.shape[0], max_voxels, replace=True)
+            cur_sweep_points = voxels[choices]
+
+        if self.prev_voxel_num > 0:
+            voxels, coors, num_points_per_voxel = \
+                self.prev_sweep_sampler.generate(prev_sweeps_points)
+            max_voxels = self.prev_voxel_num
+            if voxels.shape[0] >= max_voxels:
+                prev_sweeps_points = voxels[:max_voxels]
+            else:
+                choices = np.random.choice(
+                    voxels.shape[0], max_voxels, replace=True)
+                prev_sweeps_points = voxels[choices]
+            points = np.concatenate([cur_sweep_points, prev_sweeps_points],
+                                    0).squeeze(1)
+            results['points'] = points
+        else:
+            results['points'] = cur_sweep_points
```

Comment:
We need to also use the random choice to sample annotations in pts_masks, although we do not need it for now.